crates/blockchain-tree/src/blockchain_tree.rs

```diff
@@ -459,27 +465,58 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTree<DB, C, EF>
         Ok(())
     }
 
-    /// Same as [BlockchainTree::insert_block_with_senders] but expects that the block is in range,
-    /// See [BlockchainTree::ensure_block_is_in_range].
-    pub(crate) fn insert_in_range_block_with_senders(
-        &mut self,
-        block: SealedBlockWithSenders,
-    ) -> Result<BlockStatus, Error> {
+    /// Check if block is found inside chain and if the chain extends the canonical chain
+    /// if it does extends the canonical chain, return `BlockStatus::Valid`
+    /// if it does not extends the canonical chain, return `BlockStatus::Accepted`
+    fn is_block_inside_chain(&self, block: &BlockNumHash) -> Option<BlockStatus> {
         // check if block known and is already inside Tree
-        if let Some(chain_id) = self.block_indices.get_blocks_chain_id(&block.hash()) {
+        if let Some(chain_id) = self.block_indices.get_blocks_chain_id(&block.hash) {
             let canonical_fork = self.canonical_fork(chain_id).expect("Chain id is valid");
             // if blockchain extends canonical chain
             return if canonical_fork == self.block_indices.canonical_tip() {
-                Ok(BlockStatus::Valid)
+                Some(BlockStatus::Valid)
             } else {
-                Ok(BlockStatus::Accepted)
+                Some(BlockStatus::Accepted)
             }
         }
+        None
+    }
 
-        // check if block is part of canonical chain
-        if self.block_indices.canonical_hash(block.number) == Some(block.hash()) {
-            // block is part of canonical chain
-            return Ok(BlockStatus::Valid)
+    /// Insert a block (with senders recovered) in the tree.
+    ///
+    /// Returns the [BlockStatus] on success:
+    ///
+    /// - The block is already part of a sidechain in the tree, or
+    /// - The block is already part of the canonical chain, or
+    /// - The parent is part of a sidechain in the tree, and we can fork at this block, or
+    /// - The parent is part of the canonical chain, and we can fork at this block
+    ///
+    /// Otherwise, and error is returned, indicating that neither the block nor its parent is part
+    /// of the chain or any sidechains.
+    ///
+    /// This means that if the block becomes canonical, we need to fetch the missing blocks over
+    /// P2P.
+    ///
+    /// # Note
+    ///
+    /// If the senders have not already been recovered, call
+    /// [`BlockchainTree::insert_block_without_senders`] instead.
+    pub fn insert_block(&mut self, block: SealedBlockWithSenders) -> Result<BlockStatus, Error> {
+        self.insert_block_inner(block, true)
+    }
+
+    /// Insert a block (with senders recovered) in the tree. Check [`BlockchainTree::insert_block`]
+    /// for more info
+    pub(crate) fn insert_block_inner(
+        &mut self,
+        block: SealedBlockWithSenders,
+        do_is_known_check: bool,
```

Comment:
do we call this with `false`?

Reply:
yep, right below in `try_connect_buffered_blocks`